dvc/repo/reproduce.py

```diff
@@ -96,25 +95,23 @@ def plan_repro(
 
     The derived evaluation of _downstream_ B would be: [B, D, E]
     """
-    from .graph import get_pipeline, get_pipelines, get_steps
-
-    if pipeline or all_pipelines:
-        pipelines = get_pipelines(graph)
-        if stages and pipeline:
-            pipelines = [get_pipeline(pipelines, stage) for stage in stages]
+    import networkx as nx
 
-        leaves: List["Stage"] = []
-        for pline in pipelines:
-            leaves.extend(node for node in pline if pline.in_degree(node) == 0)
-        stages = ldistinct(leaves)
+    from .graph import get_pipeline, get_pipelines, get_steps
 
     active = _remove_frozen_stages(graph)
+    if stages and pipeline:
+        pipelines = get_pipelines(active)
+        used_pipelines = [get_pipeline(pipelines, stage) for stage in stages]
+        # create a disjointed union of all the pipelines
+        active = nx.compose_all(used_pipelines)
```

Comment:
I am building graph for all cases, except `--single-item`. This might be useful if we implement https://github.com/iterative/dvc/issues/8647 and parallel repro.

Reply:
> What was it doing before? I don't quite follow how the code changes relate to `--downstream`.

@dberenbaum, if you have a dag like following:
```mermaid
graph TD
A-->B
B-->C
```

When you do `dvc repro --pipeline B`, it'd run in order `[A, B, C]` as expected. But on `--downstream`, it would just run `C`, the last node.

DVC was searching for the leaf node of each pipeline, which in this case is `C`. By default, it finds all upstream nodes of that leaf node and runs them, which is a pipeline. 

As a side effect of this logic, on the `--downstream`, it'd try to run downstream of `C` node.
Which means it only ran leaf nodes.

---
```python
pipelines = get_pipelines(active)
pipelines = [get_pipeline(pipelines, stage) for stage in stages]
for pline in pipelines:
    leaves.extend(node for node in pline if pline.in_degree(node) == 0)
```

This was the previous code, that 1) finds all the pipelines, 2) filter pipelines that only have those nodes/stages, and 3) tries to find all the leaf node in that pipeline. (`in_degree == 0` check means that it has no arc pointing towards it, but our `repo.graph` is reversed, so take it as having no arc pointing out.)

Regarding the code changes, we don't need to compute leaf node, then find upstream nodes at all. We can walk over the pipeline directly, so we filter like before: 
```python
used_pipelines = [get_pipeline(pipelines, stage) for stage in stages]
```

The other changes here is that there may be other disconnected pipelines, which are merged together, so that we can work with a single graph. 
```python
active = nx.compose_all(used_pipelines)
```

Then, we just walk like before, but we don't pass `downstream` at all.
```python
return get_steps(active)
```